psydac/core/tests/test_kernels.py

```diff
@@ -31,133 +42,59 @@
-    if not weight.coeffs.ghost_regions_in_sync:
-        weight.coeffs.update_ghost_regions()
-
-
-    # Giving random values
-    if kind in ['hcurl', 'hdiv']:
-        for i in range(ldim):
-            field.fields[i].coeffs._data[:] = np.random.random(field.fields[i].coeffs._data.shape)
-            weight.fields[i].coeffs._data[:] = np.random.random(weight.fields[i].coeffs._data.shape)
-    else:
-        field.coeffs._data[:] = np.random.random(field.coeffs._data.shape)
-        weight.coeffs._data[:] = np.random.random(weight.coeffs._data.shape)
-
+    space_h = mapping.space
     # Preprocessing
     is_nurbs = isinstance(mapping, NurbsMapping)
-    grid = []
-    ncells = []
-    for i in range(ldim):
-        grid_i_initial = mapping.space.breaks[i]
-        ncells.append(len(grid_i_initial) - 1)
-        grid.append(np.asarray(refine_array_1d(grid_i_initial, refine, remove_duplicates=False)))
-
-    n_eval_points = [refine + 1] * ldim
-    shape_grid = tuple(grid[i].size for i in range(ldim))
-    tensor_grid = [np.reshape(grid[i], (ncells[i], n_eval_points[i])) for i in range(ldim)]
-
-    pads_m, \
-    degree_m, \
-    global_basis_m, \
-    global_spans_m = mapping.space.preprocess_regular_tensor_grid(tensor_grid, der=1)
-
-    if kind not in ['hcurl', 'hdiv']:
-        pads_s, \
-        degree_s, \
-        global_basis_s, \
-        global_spans_s = spaceh.preprocess_regular_tensor_grid(tensor_grid, der=0)
 
+    ncells = tuple(len(space_h.breaks[i]) - 1 for i in range(ldim))
+    regular_grid = [np.concatenate(
+                    [np.random.random(size=npts_per_cell) * (
+                                                             space_h.breaks[i][j + 1] 
+                                                             - space_h.breaks[i][j]
+                                                            ) 
+                                                             + space_h.breaks[i][j]
+                     for j in range(ncells[i])
+                    ]
+                    ) 
+                    for i in range(ldim)]
 
     # Direct API
     try:
         if ldim == 2:
-            jacobian_matrix_direct = np.array([[mapping.jac_mat(e1, e2) for e2 in grid[1]] for e1 in grid[0]])
+            jacobian_matrix_direct = np.array([[mapping.jac_mat(e1, e2) for e2 in regular_grid[1]] for e1 in regular_grid[0]])
 
         if ldim == 3:
             jacobian_matrix_direct = np.array([[[mapping.jac_mat(e1, e2, e3)
-                                                 for e3 in grid[2]]
-                                                for e2 in grid[1]]
-                                               for e1 in grid[0]])
+                                                 for e3 in regular_grid[2]]
+                                                for e2 in regular_grid[1]]
+                                               for e1 in regular_grid[0]])
     except NotImplementedError:
         pass
```

Comment:
The problem is more complicated than that. This tests wants to check 2 things: 
1. That the kernel functions gives the correct values (same as what I get from point-wise evaluation)
2. That the kernels functions are sensible, i.e. that the computed jacobian matrix, jacobian inverse and jacobian determinant are coherent with one another. 

1 can only be done with some arguments because `NurbsMapping` doesn't have point-wise evaluation.

Reply:
I don't know how to cleanly catch that with `pytest`.

Reply:
Understood, I will look into this

Reply:
Actually, I think that we can easily add the point-wise evaluation algorithms to `NurbsMapping`. The missing piece is in function [`TensorFemSpace.eval_field_gradient`](https://github.com/pyccel/psydac/blob/084e2996c11fac8650fb1751306ad3140207d1d2/psydac/fem/tensor.py#L616), which accepts an optional `weight` argument but does not use it. Once that is implemented, we can easily implement the function [`NurbsMapping.jac_mat`](https://github.com/pyccel/psydac/blob/084e2996c11fac8650fb1751306ad3140207d1d2/psydac/mapping/discrete.py#L934).

Reply:
Done in [745be55](https://github.com/pyccel/psydac/pull/214/commits/745be5502de8969f589ef5f03d080217f748eca4)